lib/kino/smart_cell/chart_builder.ex

```diff
@@ -0,0 +1,180 @@
+      "width" => attrs["width"],
+      "height" => attrs["height"],
+      "x_axis" => attrs["x_axis"],
+      "y_axis" => attrs["y_axis"],
+      "x_axis_type" => attrs["x_axis_type"],
+      "y_axis_type" => attrs["y_axis_type"],
+      "color" => attrs["color"],
+      "color_type" => attrs["color_type"]
+    }
+
+    {:ok, assign(ctx, fields: fields, options: %{}, vl_alias: nil, missing_dep: missing_dep())}
+  end
+
+  @impl true
+  def scan_binding(pid, binding, env) do
+    send(pid, {:scan_binding_result, binding, env})
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    payload = %{
+      fields: ctx.assigns.fields,
+      missing_dep: ctx.assigns.missing_dep,
+      options: ctx.assigns.options
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_info({:scan_binding_result, binding, env}, ctx) do
+    dfs = Keyword.filter(binding, fn {_key, val} -> is_map(val) end)
+    data_options = Enum.map(dfs, fn {k, v} -> {k, Map.keys(v)} end) |> Enum.into(%{})
+    ctx = assign(ctx, options: data_options, vl_alias: vl_alias(env))
+
+    broadcast_event(ctx, "set_data_options", %{"options" => data_options})
+
+    {:noreply, ctx}
+  end
+
+  @impl true
+  def handle_event("update_field", %{"field" => field, "value" => value}, ctx) do
+    updated_fields = to_updates(ctx.assigns.fields, field, value)
+    ctx = update(ctx, :fields, &Map.merge(&1, updated_fields))
+    if field == "data", do: update_options(ctx, value)
+
+    broadcast_event(ctx, "update", %{"fields" => updated_fields})
+
+    {:noreply, ctx}
+  end
+
+  defp update_options(ctx, df) do
+    df = String.to_atom(df)
+    options = ctx.assigns.options[df]
+    broadcast_event(ctx, "set_axis_options", %{"options" => options})
+  end
+
+  defp to_updates(_fields, field, ""), do: %{field => nil}
+
+  defp to_updates(_fields, field, value) when field in @as_int do
+    %{field => String.to_integer(value)}
+  end
+
+  defp to_updates(_fields, field, value) when field in @as_atom do
+    %{field => String.to_atom(value)}
+  end
+
+  defp to_updates(_fields, field, value), do: %{field => value}
+
+  defp vl_alias(%Macro.Env{aliases: [{vl_alias, VegaLite}]}), do: vl_alias
+  defp vl_alias(_), do: nil
+
+  @impl true
+  def to_attrs(ctx) do
+    ctx.assigns.fields
+    |> Map.put("vl_alias", ctx.assigns.vl_alias)
+  end
+
+  @impl true
+  def to_source(attrs) do
+    attrs
+    |> to_quoted()
+    |> Macro.to_string()
+    |> Code.format_string!()
+    |> IO.iodata_to_binary()
+  end
+
+  defp to_quoted(%{"data" => nil}), do: nil
+
+  defp to_quoted(attrs) do
+    data = attrs["data"]
+    module = if attrs["vl_alias"], do: attrs["vl_alias"], else: VegaLite
+    module = Module.split(module) |> hd() |> String.to_atom()
+
+    attrs =
+      attrs
+      |> Enum.map(fn {k, v} -> {String.to_atom(k), v} end)
+      |> Enum.into(%{})
+
+    [{_, root} | nodes] = [
+      new: %{name: :new, module: module, args: [[width: attrs.width, height: attrs.height]]},
```

Comment:
```suggestion
      %{field: :new, function_name: :new, module: module, args: [[width: attrs.width, height: attrs.height]]},
```

I was thinking about this data structure, and now it seems to me that abandoning the outer keyword in favor of a list of maps could be better.

Some matches below would also need to change, but it makes for a cleaner code overall. `[root | nodes] = ...` for instance